src/map.cpp

```diff
@@ -398,85 +445,72 @@ const vehicle *map::vehproceed()
     }
 
     // Low enough for bicycles to go in reverse.
-    if( abs( veh.velocity ) < 20 ) {
+    if( !should_fall && abs( veh.velocity ) < 20 ) {
         veh.stop();
     }
 
-    if( veh.velocity == 0 ) {
+    if( !should_fall && abs( veh.velocity ) < 20 ) {
         veh.of_turn -= .321f;
         return &veh;
     }
 
-    // TODO: Handle wheels in the air
-    std::vector<int> float_indices = veh.all_parts_with_feature(VPFLAG_FLOATS, false);
-    if( float_indices.empty() ) {
-        // sink in water?
-        std::vector<int> wheel_indices = veh.all_parts_with_feature(VPFLAG_WHEEL, false);
-        int num_wheels = wheel_indices.size();
-        int submerged_wheels = 0;
-        for (int w = 0; w < num_wheels; w++) {
-            const int p = wheel_indices[w];
-            const tripoint pp = pt + veh.parts[p].precalc[0];
-            // deep water
-            if (ter_at(pp).has_flag(TFLAG_DEEP_WATER)) {
-                submerged_wheels++;
-            }
+    const float traction = vehicle_traction( veh );
+    // TODO: Remove this hack, have vehicle sink a z-level
+    if( traction < 0 ) {
+        add_msg(m_bad, _("Your %s sank."), veh.name.c_str());
+        if( pl_ctrl ) {
+            veh.unboard_all();
         }
-        // submerged wheels threshold is 2/3.
-        if( num_wheels > 0 && (float)submerged_wheels / num_wheels > .666) {
-            add_msg(m_bad, _("Your %s sank."), veh.name.c_str());
-            if( pl_ctrl ) {
-                veh.unboard_all();
-            }
-            if( g->remoteveh() == &veh ) {
-                g->setremoteveh( nullptr );
-            }
-            // destroy vehicle (sank to nowhere)
-            destroy_vehicle( &veh );
-            // The returned pointer is always a dangling one!
-            return &veh;
+        if( g->remoteveh() == &veh ) {
+            g->setremoteveh( nullptr );
         }
-    } else {
-        int num = float_indices.size(), moored = 0;
-        for( int w = 0; w < num; w++ ) {
-            const int p = float_indices[w];
-            const tripoint pp = pt + veh.parts[p].precalc[0];
-
-            if( !has_flag( "SWIMMABLE", pp ) ) {
-                moored++;
-            }
-        }
-
-        if( moored > num - 1 ) {
+        // destroy vehicle (sank to nowhere)
+        destroy_vehicle( &veh );
+        // The returned pointer is always a dangling one!
```

Comment:
Whoa, stop right there, what's that? You're returning an invalid *non-null* pointer on purpose? The caller has no way to distinguish it from a valid non-null pointer. And I see the returned pointer being used:
```C++
        const vehicle *veh = vehproceed();
        if( veh == nullptr ) {
            break;
        } else {
            on_vehicle_moved( veh->smz ); // Hello undefined behavior.
        }
```